### PR TITLE
[WIP] MSCXXXX: `matrix.to` URI syntax v2

### PR DESCRIPTION
**Pull request by jryans**
[Rendered](https://github.com/matrix-org/matrix-doc/blob/jryans/matrix-to-uri-v2/proposals/0000-matrix.to-v2.md)

### CODE REVIEW

File: proposals/2644-matrix.to-v2.md [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+only the event ID in the `identifier` position for new links.
+
+| type | literal value | encoded value |
+| ---- | ------------- | ------------- |
+| event ID | `$event:example.org` | `%24event%3Aexample.org` |
+
+Since clients currently cannot find a room from the event ID alone, a new
+client-server API is added to support the new format with only an event ID.
+
+> TODO: To support this, a new client-server API will be defined which
```
Comment (jryans) [inline proposals/2644-matrix.to-v2.md:90]: This is the main reason I have marked this WIP for now, so please let me know how best to proceed for this portion.

Comment (turt2live) [reply proposals/2644-matrix.to-v2.md:90]: oh, hmm, if this is the only reason then it might be best to stick it in the general not-WIP queue for more eyes.

So far our answer to scope creep like this is to prefer a new MSC.

File: proposals/2644-matrix.to-v2.md [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+of a permalink that references a specific event, as shown in the table below.
+Going forward, this should be considered deprecated, and clients should place
+only the event ID in the `identifier` position for new links.
+
+| type | literal value | encoded value |
+| ---- | ------------- | ------------- |
+| event ID | `$event:example.org` | `%24event%3Aexample.org` |
+
+Since clients currently cannot find a room from the event ID alone, a new
+client-server API is added to support the new format with only an event ID.
```
Comment (t3chguy) [inline proposals/2644-matrix.to-v2.md:88]: How would this work for v1 or v2 rooms where the Event ID is not based on a hash. Even if post-v3 world hash-event-ids could clash. https://matrix.org/docs/spec/rooms/v3#event-ids

Comment (tulir) [reply proposals/2644-matrix.to-v2.md:88]: The endpoint already exists, but is deprecated: https://matrix.org/docs/spec/client_server/latest#deprecated-get-matrix-client-r0-events-eventid

It was presumably deprecated for a reason though

File: proposals/2644-matrix.to-v2.md (hunk 1 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:90]: Okay, will remove the WIP label then.

File: proposals/2644-matrix.to-v2.md (hunk 2 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:88]: Ah yes, sorry I meant to state that only v3 room and later event IDs can be used in this way. I'll update the text to reflect that.

Comment (t3chguy) [reply proposals/2644-matrix.to-v2.md:88]: Arguably it should be v4 to be url-safe: https://matrix.org/docs/spec/rooms/v4

Comment (turt2live) [reply proposals/2644-matrix.to-v2.md:88]: URL safety is a thing that's supposed to be in the existing spec, but none of the clients (literally none) handle it correctly.

Comment (jryans) [reply proposals/2644-matrix.to-v2.md:88]: About event IDs clashing, if that ends up happening, I assume that would be a much larger problem for other parts of the system as well, so I don't think it requires special consideration (like carrying around a room ID just in case) here...?

Comment (turt2live) [reply proposals/2644-matrix.to-v2.md:88]: Yesterday I found out synapse has a UNIQUE constraint on the event ID in various tables (like `events`), so it shouldn't happen that often.

File: proposals/2644-matrix.to-v2.md [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+* `sharer=<MXID>`
+  * This parameter allows indicating the MXID of the account which created the
+    link, so that clients and interstitial UIs can display more context to the
+    user
+
+If multiple ``<additional arguments>`` are present, they should be joined by `&`
+characters, as in `https://matrix.to/!somewhere%3Aexample.org?via=example.org&via=alt.example.org`
+
+The components of the matrix.to URI (``<identifier>`` and ``<extra parameter>``)
+are to be percent-encoded as per RFC 3986.
```
Comment (KitsuneRal) [inline proposals/2644-matrix.to-v2.md:122]: See my comment above, except that the proposed `client` parameter has an embedded URL in it which may include queries and even fragments; so it becomes pretty important to clearly delineate which delimiters are encoded (at least those in the client URL, obviously) and which are not (`?` and `/` outside of the URL? then we're insisting on non-compliance with RFC3986; or, if we encode them outside the URL then `?` and `/` inside the client URL should be double-encoded - `%253F` and `%252F`, respectively?).

In other words - we've allowed some mess with non-encoding things for the sake of matrix.to usability, we now have to decide if we want to continue allowing it (with the risk of problems in apps unaware of our quirks, and client developers getting lost in the meaning of a particular character in a particular position when dealing with matrix.to URLs) or we want to get stricter (at the expense of reduced usability for humans who construct matrix.to URLs themselves).

File: proposals/2644-matrix.to-v2.md [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+Examples of matrix.to URIs using the revised syntax are:
+
+* Room alias: ``https://matrix.to/#/%23somewhere%3Aexample.org``
+* Room alias with client and sharer:
+  ``https://matrix.to/#/%23somewhere%3Aexample.org?client=foo.com&sharer=%40alice%3Aexample.org``
```
Comment (KitsuneRal) [inline proposals/2644-matrix.to-v2.md:128]: Can we give a bit less simplistic example for the client URL here? Say, if the download URL is `https://example.org/releases?download#latest`:
```suggestion
  ``https://matrix.to/#/%23somewhere%3Aexample.org?client=example.org%2Freleases%3Fdownload%23latest&sharer=%40alice%3Aexample.org``
```

File: proposals/2644-matrix.to-v2.md [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+* Room: ``https://matrix.to/#/!somewhere%3Aexample.org?via=example.org&via=alt.example.org``
```
Comment (KitsuneRal) [inline proposals/2644-matrix.to-v2.md:129]:
```suggestion
* Room: ``https://matrix.to/#/%21somewhere%3Aexample.org?via=example.org&via=alt.example.org``
```
Or become consistent on what we do with sigils and colons inside Matrix ids: either we encode them all, or we only encode `#` because the RFC doesn't allow to leave it unencoded.

File: proposals/2644-matrix.to-v2.md [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+* Event permalink: ``https://matrix.to/#/%24event%3Aexample.org?via=example.org&via=alt.example.org``
+* User: ``https://matrix.to/#/%40alice%3Aexample.org``
+* Group: ``https://matrix.to/#/%2Bexample%3Aexample.org``
+
+### Summary of changes
+
+* When permalinking to a specific event, the room ID is no longer required and
+  event IDs are now permitted in the identifier position, so URIs like
+  `https://matrix.to/#/%24event%3Aexample.org` are now acceptable
```
Comment (KitsuneRal) [inline proposals/2644-matrix.to-v2.md:138]:
```suggestion
  `https://matrix.to/#/%24EventIdAsUnpaddedBase64` are now acceptable
```

File: proposals/2644-matrix.to-v2.md [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+* Clients should prefer creating URIs with room aliases instead of room IDs
+  where possible, as it's more human-friendly and `via` params are not needed
```
Comment (KitsuneRal) [inline proposals/2644-matrix.to-v2.md:140]: That recommendation doesn't match the feature you introduced just before. First, the recommendation itself is, unfortunately, not universal - see https://github.com/vector-im/riot-web/issues/2925 for details. Second, if you link to an event with `$`, a random federating server is unlikely to find its room if it doesn't already know that room; so you'll have to pass `via` in the same way you do it for room ids. I'd probably just drop this recommendation because it's outside of this MSC's scope.

File: proposals/2644-matrix.to-v2.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+* A new, optional `client` parameter allows clients to indicate
+  which client shared the URI
+* A new, optional `federated` parameter allows indicating whether a room exists
+  on a federating server (assumed to be the default), or if the client must
+  connect via the server identified by the room ID or event ID (when set to
+  `false`)
```
Comment (KitsuneRal) [inline proposals/2644-matrix.to-v2.md:146]:
```suggestion
  on a federating server (assumed to be the default), or if it's only
  reachable through/at a certain server
```

File: proposals/2644-matrix.to-v2.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+moves forward, the Matrix URI scheme would likely need to be reworked to
+accommodate the additions here.
+
+The new `client` parameter implies there are potentially many identifiers that
+might be passed that point to a given client. If there are use cases which rely
+on a static mapping of client identifier to client name, logo, etc. for some
+reason, then that could become a burden to maintain over time. The flexibility
+of accepting any URL as an identifier (and thus avoiding the need to register a
+client in a centralised place) seems preferable and hopefully outweighs this
+concern.
```
Comment (KitsuneRal) [inline proposals/2644-matrix.to-v2.md:164]: Maybe using a rev-DNS naming as practised in many other areas throughout Matrix would still be better. With the level of maturity across the ecosystem, clients' download locations change frequently enough to become stale in permalinks.

A bigger issue though is that `client` and `sharer` set an unusual (and questionable) precedent where a Uniform Resource Locator describes not only the target location but also the locator's author. If uncontrolled, it can be a means for massive leaking of data people probably didn't want to share - specifically the connectivity of people across the social graph (imagine a reaction upon receiving a link with the _original_ sharer because the next sharer didn't care to update it: "wait, where does he know _this_ sharer from?"). And rewriting the sharing source every time the link is re-shared is really tedious; as soon as the link is copied as plain text, you can _assume_ rewriting _won't_ happen by default.

File: proposals/2644-matrix.to-v2.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+extend the future [Matrix URI
+scheme](https://github.com/matrix-org/matrix-doc/pull/2312). This proposal
+attempts to be pragmatic and tries to extend what is already in use today,
+rather than blocking on a new scheme.
+
+## Security considerations
+
+The new `sharer` parameter is not authenticated, so you could make it appear as
+if someone had shared something they did not. It is currently assumed that this
+is a minor concern.
```
Comment (KitsuneRal) [inline proposals/2644-matrix.to-v2.md:178]: Right, you can (at least in theory) distrust `sharer` but, as mentioned above, you cannot "be forced to forget" if you received a certain `sharer` string. If this piece of information leaked, it's leaked.

File: proposals/2644-matrix.to-v2.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+  * This parameter allows clients to indicate which client shared the URI
+  * Clients should identify themselves via a schemeless `https` URL pointing
+      to a download / install page, such as:
+    * `foo.com`
+    * `apps.apple.com/app/foo/id1234`
+    * `play.google.com/store/apps/details?id=com.foo.client`
+* `federated=false`
+  * This parameter allows indicating whether a room exists on a federating
+    server (assumed to be the default), or if the client must connect via the
+    server identified by the room ID or event ID (when set to `false`)
```
Comment (KitsuneRal) [inline proposals/2644-matrix.to-v2.md:112]: By the definition of federation, you cannot rely on federating servers to identify a server of a room/event residing outside of federation. Seems that `federated=false` implies using `via=<non-federating server>` except quite specific situations. The next issue is with the requirement to connect via that server: worded in almost the same way in #2312, this approach [faced controversy](https://github.com/matrix-org/matrix-doc/pull/2312#discussion_r348960282). To alleviate that, I propose the following: 
```suggestion
    server (assumed to be the default), or if the room/event is only reachable
    through a specific server (when set to `false`). This server is usually
    specified in a `via` parameter, unless the target client is expected to
    be able to unambiguously resolve it.
```

File: proposals/2644-matrix.to-v2.md [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+  * One or more servers [should be
+    specified](https://matrix.org/docs/spec/appendices#routing) in the format
+    `example.org` when linking to a room (including a permalink to an event in a
+    room) since room IDs are not currently routable
+
+If multiple ``<additional arguments>`` are present, they should be joined by `&`
+characters, as in `https://matrix.to/!somewhere%3Aexample.org?via=example.org&via=alt.example.org`
+
+The components of the matrix.to URI (``<identifier>`` and ``<extra parameter>``)
+are to be percent-encoded as per RFC 3986.
```
[the remark this answers is not in the document]
Comment (deepbluev7) [reply proposals/2644-matrix.to-v2.md:56]: I agree with this. Having to only percent-encode delimiters when necessary, but accepting both makes it much easier to write matrix.to links by hand as well as generate them programmatically. Also, sparing only ! is super weird, but the current spec does that a lot, where only some things get encoded. It may make sense to say `/`, `?` and `#` must always be encoded to accommodate webclients, but anything else should also be accepted unencoded imo.

File: proposals/2644-matrix.to-v2.md (hunk 2 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:88]: It's not clear to me that we _must_ settle whether event IDs are unique for the scope of this MSC though. The meaning of an event ID by itself as proposed here seems clear enough: a client take that event ID and map to a room using the TBD API. For room versions where the event IDs are not globally unique, you'll probably get some very strange answers back, but the description of the event to room mapping API could explain that uniqueness depends on room versions.

Comment (KitsuneRal) [reply proposals/2644-matrix.to-v2.md:88]: Well, you can say that trying to resolve a bare event URL in the presence of room versions that don't guarantee event id uniqueness leads to [undefined behaviour](https://en.m.wikipedia.org/wiki/Undefined_behavior); however, since this is exactly the situation in the current federation due to v1 and v2 rooms, such disclaimer basically says that practically _any_ attempt to resolve a bare event id is not guaranteed to return a non-conflicting result. The main issue here is: before resolving the event id you have no idea about the version of its room; so you cannot rely on any invariants specific to certain room versions.

File: proposals/2644-matrix.to-v2.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:56]: There seems to be general agreement here and in the spec room that it would be better to focus on the unencoded form as the canonical syntax, and then note that they _can_ be encoded and will be accepted in that form as well by clients and the matrix.to site itself.

File: proposals/2644-matrix.to-v2.md [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+  * Clients should identify themselves via a schemeless `https` URL pointing
+      to a download / install page, such as:
+    * `foo.com`
+    * `apps.apple.com/app/foo/id1234`
+    * `play.google.com/store/apps/details?id=com.foo.client`
+* `federated=false`
+  * This parameter allows indicating whether a room exists on a federating
+    server (assumed to be the default), or if the client must connect via the
+    server identified by the room ID or event ID (when set to `false`)
+* `sharer=<MXID>`
```
[the remark this answers is not in the document]
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:113]: The reason for an MXID in particular is it would allow matrix.to to request profile info (avatar, display name) for the sharer and display that (assuming the user consents to such requests at all). I'll fold this extra info into the text.

`Origin` seems a bit technical and hides that it's a person who shared a thing, so I'm inclined to leave the name as-is. I also worry it would be easily confused with the various servers and vias flying around in the syntax.

File: proposals/2644-matrix.to-v2.md [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+| type | literal value | encoded value |
+| ---- | ------------- | ------------- |
+| event ID | `$event:example.org` | `%24event%3Aexample.org` |
+
+Since clients currently cannot find a room from the event ID alone, a new
+client-server API is added to support the new format with only an event ID.
+
+> TODO: To support this, a new client-server API will be defined which
+> allows clients to query the mapping from event ID to room ID. I'd like
+> feedback on whether to define such a thing here, or in a separate MSC.
```
[the remark this answers is not in the document]
Comment (jryans) [reply proposals/2644-matrix.to-v2.md]: Agreed this should happen separately, updated the text for now to state that.

File: proposals/2644-matrix.to-v2.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:113]: Added explanation for MXID to the text, hopefully that clarifies.

[the remark this answers is not in the document]
Comment (KitsuneRal) [reply proposals/2644-matrix.to-v2.md:113]: Well, I still wouldn't mind to have the same feature for non-Matrix sharers but if we want to make some kind of experience barrier between Matrix and non-Matrix then so be it.

[the remark this answers is not in the document]
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:113]: Ah, I see what you mean... One possibility is to allow it to be more free form and generic, but also suggest MXIDs as the thing to put here if you have one. I'm curious what others thinks about this as well.

File: proposals/2644-matrix.to-v2.md [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+    specified](https://matrix.org/docs/spec/appendices#routing) in the format
+    `example.org` when linking to a room or an event since room IDs are
+    not currently routable
+* `client=<client URL>`
+  * This parameter allows clients to indicate which client shared the URI
+  * Clients should identify themselves via a schemeless `https` URL pointing
+      to a download / install page, such as:
+    * `foo.com`
+    * `apps.apple.com/app/foo/id1234`
+    * `play.google.com/store/apps/details?id=com.foo.client`
```
Comment (t3chguy) [inline proposals/2644-matrix.to-v2.md:108]: M suggested that matrix.to should scrape the Name & Icon from this to show to the user but play.google.com doesn't offer ACAO so wouldn't be possible without some server-side service.

File: proposals/2644-matrix.to-v2.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:56]: To give a bit more detail, the encoding rules in this MSC at the moment are simply copied from the [current spec text](https://matrix.org/docs/spec/appendices#matrix-to-navigation): it encodes `#`, `+`, `$`, but does not encode `!`. (Originally I just assumed that encoding did indeed match the RFC as well, since spec text says it should.)

Anyway, I plan to change all the identifiers to be unencoded, since that is what's in common use, and obviously nicer for humans to read.

File: proposals/2644-matrix.to-v2.md (hunk 2 above)
Comment (turt2live) [reply proposals/2644-matrix.to-v2.md:88]: > Ah yes, sorry I meant to state that only v3 room and later event IDs can be used in this way. I'll update the text to reflect that.

Is it possible to have this happen? We probably shouldn't call out specific versions (to account for future mistakes), and could easily say something like "matrix.to URIs using a bare event ID should never be generated when the global uniqueness of the event cannot be guaranteed, such as in the case of v1 and v2 rooms."

File: proposals/2644-matrix.to-v2.md [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+> defined in a separate MSC.
+
+The ``<additional arguments>`` and the preceding question mark are optional and
+only apply in certain circumstances:
+
+* `via=<server>`
+  * One or more servers [should be
+    specified](https://matrix.org/docs/spec/appendices#routing) in the format
+    `example.org` when linking to a room or an event since room IDs are
+    not currently routable
```
Comment (turt2live) [inline proposals/2644-matrix.to-v2.md:101]:
```suggestion
    `example.org` when linking to a room or an event since room IDs and event
    IDs are not currently routable on their own
```

File: proposals/2644-matrix.to-v2.md [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+* `client=<client URL>`
+  * This parameter allows clients to indicate which client shared the URI
```
Comment (turt2live) [inline proposals/2644-matrix.to-v2.md:103]: would be good to spell out that this also gives users the option of downloading the client, particularly for use in the matrix.to fallback page (which is specifically specced to not exist, ironically).

File: proposals/2644-matrix.to-v2.md [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+  * Clients should identify themselves via a schemeless `https` URL pointing
+      to a download / install page, such as:
+    * `foo.com`
+    * `apps.apple.com/app/foo/id1234`
+    * `play.google.com/store/apps/details?id=com.foo.client`
+* `federated=false`
```
Comment (turt2live) [inline proposals/2644-matrix.to-v2.md:109]: Why do we need this? `via` indeed appears to communicate the intent, and if a client is using Matrix in a private environment it is going to be better off inventing its own permalink structure (for in-app linking rather than trying to tame matrix.to)

File: proposals/2644-matrix.to-v2.md [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+  * This parameter allows indicating whether a room exists on a federating
+    server (assumed to be the default), or if the client must connect via the
+    server identified by the room ID or event ID (when set to `false`)
+* `sharer=<MXID>`
```
Comment (turt2live) [inline proposals/2644-matrix.to-v2.md:113]: Do we have any abuse mitigation suggestions to provide to clients which handle matrix.to in-app (or to the matrix.to fallback page)? It's far too easy to manually craft this to use non-existent IDs which will easily get clients to show in giant header text the bare mxid. Or worse, a real mxid with all kinds of abuse being included in the profile response.

File: proposals/2644-matrix.to-v2.md [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+  * This parameter allows indicating the MXID of the account which created the
+    link, so that clients and interstitial UIs can display more context to the
+    user
+  * As an example, clients and interstitial UIs could use this to query profile
+    data for the sharer's account and display the sharer's avatar and display
+    name
+
+If multiple ``<additional arguments>`` are present, they should be joined by `&`
+characters, as in `https://matrix.to/!somewhere%3Aexample.org?via=example.org&via=alt.example.org`
```
Comment (turt2live) [inline proposals/2644-matrix.to-v2.md:122]: Can we just say it's query string format under RFC whatever?

File: proposals/2644-matrix.to-v2.md [hunk 21]
```diff
@@ -0,0 +1,187 @@
+# MSC2644: `matrix.to` URI syntax v2
+
+`matrix.to` URIs are used frequently to link to Matrix things (events, rooms,
+users, groups) in the ecosystem today. By adjusting and extending them a bit
+further, both clients and the interstitial screens hosted at `matrix.to` can
+give the user more context and a better experience.
```
Comment (turt2live) [inline proposals/2644-matrix.to-v2.md:6]: it would be good to acknowledge and re-affirm the spec by saying that the web service is not guaranteed to exist, work, etc. Most of these changes are to better support that interface, however the client benefits of being able to iterate on their own consumption of these links shouldn't be missed.

File: proposals/2644-matrix.to-v2.md (hunk 20 above)
Comment (KitsuneRal) [reply proposals/2644-matrix.to-v2.md:122]: It would at least need the "as-if we were not inside the fragment already" caveat.

File: proposals/2644-matrix.to-v2.md (hunk 11 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:112]: Hmm, why would the `federated=false` case include `via=<non-federating server>`? I had been assuming for the non-federated case, the client effectively must connect to the only server that knows about the room, which is already present in the room ID, so there's not much purpose to including `via` as well for such a case. Is "unless the target client is expected to be able to unambiguously resolve it" trying to allow for this somehow...? It's vague enough that I am not sure what it really means... 😅 

The goal I see with `federated=false` is it allows the client to know for certain that it must connect via an account on the only server that knows about that room, rather than attempting to do so using `via` params that will not work from a federating account on some different server than the room.

File: proposals/2644-matrix.to-v2.md (hunk 18 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:109]: The goal I see with `federated=false` is it allows the client to know for certain that it must connect via an account on the only server that knows about that room, rather than attempting to do so using via params that will not work from a federating account on some different server than the room.

Comment (turt2live) [reply proposals/2644-matrix.to-v2.md:109]: From a spec perspective I'm not sure we care about that. If the user received a link, they have a reasonable shot of going through to the content.

File: proposals/2644-matrix.to-v2.md (hunk 11 above)
Comment (KitsuneRal) [reply proposals/2644-matrix.to-v2.md:112]: Because a client can be exposed to more than one Matrix network? Actually, it just occurred to me that `federated=false` is not quite a good name for that reason. If you have one public federation and one closed federation (for testing, e.g.) then you actually have two federations, not one federation and an unfederated server :-\

And continuing to the point about using room IDs to find an unfederated server - with certain confidence this works if you have a single entirely unfederated server _and_ it has never federated. But IIRC, Synapse already adds the server signified in the room id to the list of "via" servers, and we even have this behaviour specced somewhere - so probably `federated=false` is just extraneous then?

The "able to unambiguously resolve" piece is, yes, an attempt to describe the situation when the link is shared with someone who _most definitely_ will use it at the correct server. Feel free to rephrase :-D

File: proposals/2644-matrix.to-v2.md (hunk 18 above)
Comment (KitsuneRal) [reply proposals/2644-matrix.to-v2.md:109]: The author of the link may or may not supply `via` parts as necessary, basing on how the room is supposed to be reached. Since unfederated rooms by definition can only have unfederated accounts, the `via` list will contain exactly one entry - the unfederated server (if [the current algorithm](https://matrix.org/docs/spec/appendices#routing) of constructing it is followed).

File: proposals/2644-matrix.to-v2.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (jryans) [reply proposals/2644-matrix.to-v2.md]: I have updated the MSC to generally not percent-encode, aside from the new `client` argument, which itself contains a URL, so needs some encoding. For any new comments on the revised text, let's start a new thread.

File: proposals/2644-matrix.to-v2.md (hunk 3 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md]: I have updated the MSC to generally not percent-encode, aside from the new `client` argument, which itself contains a URL, so needs some encoding. For any new comments on the revised text, let's start a new thread.

File: proposals/2644-matrix.to-v2.md (hunk 5 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md]: I have updated the MSC to generally not percent-encode, aside from the new `client` argument, which itself contains a URL, so needs some encoding. For any new comments on the revised text, let's start a new thread.

File: proposals/2644-matrix.to-v2.md (hunk 21 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:6]: Okay, I have added a bit more text on this.

File: proposals/2644-matrix.to-v2.md (hunk 20 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md]: The description of fragment in RFC 3986 seems quite lenient about which characters can be used. Why is caveat needed?

Comment (jryans) [reply proposals/2644-matrix.to-v2.md]: I have at least added a mention of the RFC here.

File: proposals/2644-matrix.to-v2.md (hunk 19 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md:126]: Hmm, I am not sure what we might offer in that regard... In seems like more of social problem to me, as in you should avoid clicking links from untrusted parties who might abuse things in this way. I can't think of technical solution, other than somehow signing the links, but that's quite heavy weight for this problem and aesthetically unsatisfying.

File: proposals/2644-matrix.to-v2.md (hunk 6 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md]: Thanks, I have updated to use v4+ style event IDs by default.

File: proposals/2644-matrix.to-v2.md [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+
+```
+https://matrix.to/#/<identifier>/<extra parameter>?<additional arguments>
+```
+
+The `identifier` (required) may be a:
+
+| type | value |
+| ---- | ----- |
+| event ID (v1) | `$event:example.org` |
```
Comment (Sorunome) [inline proposals/2644-matrix.to-v2.md:81]: As event IDs should be per-room unique they shouldn't be able to be placed on their own in the URL, only in combination with a room ID.

Comment (Sorunome) [reply proposals/2644-matrix.to-v2.md]: While soru highly disagrees with the update, as then in the future the format will be that only in e.g. v3-v9 rooms the event ID should be alone, the discussion for that should be moved for consistency here: https://github.com/matrix-org/matrix-doc/pull/2695#discussion_r459641368

File: proposals/2644-matrix.to-v2.md (hunk 2 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md]: I have updated the text to state event IDs alone should only be used when global uniqueness can be guaranteed, as @turt2live suggests, which I believe resolves the core concern of this thread. Let's start a new thread if there are other details remaining here.

Comment (Sorunome) [reply proposals/2644-matrix.to-v2.md]: There's also discussion on this here: https://github.com/matrix-org/matrix-doc/pull/2695#discussion_r459641368

File: proposals/2644-matrix.to-v2.md (hunk 22 above)
Comment (jryans) [reply proposals/2644-matrix.to-v2.md]: Thanks, I have revised the text as per an earlier thread here to say event IDs should only be used alone when uniqueness is assured.

Comment (Sorunome) [reply proposals/2644-matrix.to-v2.md]: Unfortunatley that means that in the future it may have to be updated that it only applies to e.g. room version 3-9 and then 12-14. This is a huge mess :/

Comment (jryans) [reply proposals/2644-matrix.to-v2.md]: As you suggest, I think the thread on 2695 is indeed the better place to discuss and the global uniqueness issue.

**Comment by richvdh** [comment]
What is the state of play with this MSC? Does it need updating to reflect the decisions made for `matrix:` URIs in [MSC2312](https://github.com/matrix-org/matrix-doc/pull/2312) ?  Does it in fact want to be parked and superceded by `matrix:` URIs?

**Comment by jryans** [comment]
> What is the state of play with this MSC? Does it need updating to reflect the decisions made for `matrix:` URIs in [MSC2312](https://github.com/matrix-org/matrix-doc/pull/2312) ? Does it in fact want to be parked and superceded by `matrix:` URIs?

My understanding is this got a bit lost in the sands of time when we rewrote matrix.to, which went through a series of several approaches. I'm checking with @bwindels in https://github.com/matrix-org/matrix.to/issues/205#issuecomment-858770669 to see if any of this was ever implemented in the end.

If not, it may be better at this point to discard this and let `matrix:` URIs lead the way forward.

**Comment by anoadragon453** [comment]
Looking at https://github.com/matrix-org/matrix.to/issues/205#issuecomment-858770669, it appears an implementation has not happened. Does that spell the end of the road for this MSC?

**Comment by jryans** [comment]
This MSC is not actively being worked on... Perhaps it should be closed? Feel free to do so if that's reasonable.
